core/services/workflows/engine.go

```diff
@@ -505,86 +532,71 @@ func (e *Engine) startExecution(ctx context.Context, executionID string, event *
 		return err
 	}
 
+	ch := make(chan store.WorkflowExecutionStep)
+	added := e.stepUpdatesChMap.add(executionID, stepUpdateChannel{
+		ch:          ch,
+		executionID: executionID,
+	})
+	if !added {
+		// skip this execution since there's already a stepUpdateLoop running for the execution ID
+		e.logger.With(eIDKey, executionID).Debugf("won't start execution for execution %s, execution was already started", executionID)
+		return nil
+	}
+	e.wg.Add(1)
+	go e.stepUpdateLoop(ctx, executionID, ch, dbWex.CreatedAt)
+
 	for _, td := range triggerDependents {
 		e.queueIfReady(*ec, td)
 	}
 
 	return nil
 }
 
-func (e *Engine) handleStepUpdate(ctx context.Context, stepUpdate store.WorkflowExecutionStep) error {
+func (e *Engine) handleStepUpdate(ctx context.Context, stepUpdate store.WorkflowExecutionStep, workflowCreatedAt *time.Time) error {
+	l := e.logger.With(eIDKey, stepUpdate.ExecutionID, sRKey, stepUpdate.Ref)
+
+	// If we've been executing for too long, let's time the workflow out and stop here.
+	if workflowCreatedAt != nil && e.clock.Since(*workflowCreatedAt) > e.maxExecutionDuration {
+		l.Info("execution timed out")
+		return e.finishExecution(ctx, stepUpdate.ExecutionID, store.StatusTimeout)
```

Comment:
@vyzaldysanchez This should update the state of the step, and then do the normal propagation logic that we do down below.